cpu/stm32_common/periph/usbdev.c

```diff
@@ -0,0 +1,1151 @@
+}
+
+static void _set_mode_device(stm32_usb_otg_fshs_t *usbdev)
+{
+    const stm32_usb_otg_fshs_config_t *conf = usbdev->config;
+
+    /* Force device mode */
+    _global_regs(conf)->GUSBCFG |= USB_OTG_GUSBCFG_FDMOD;
+    /* Spinlock to prevent a context switch here, needs a delay of 25 ms when
+     * force switching mode */
+    xtimer_spin(xtimer_ticks_from_usec(25 * US_PER_MS));
+}
+
+static void _usbdev_init(usbdev_t *dev)
+{
+    /* Block both STOP and STANDBY, STOP is unblocked during USB suspend
+     * status */
+    pm_block(STM32_PM_STOP);
+    pm_block(STM32_PM_STANDBY);
+
+    stm32_usb_otg_fshs_t *usbdev = (stm32_usb_otg_fshs_t *)dev;
+    const stm32_usb_otg_fshs_config_t *conf = usbdev->config;
+
+    /* Enable the clock to the peripheral */
+    periph_clk_en(conf->ahb, conf->rcc_mask);
+
+    _enable_gpio(conf);
+
+    /* TODO: implement ULPI mode when a board is available */
+#ifdef STM32_USB_OTG_HS_ENABLED
+    if (conf->type == STM32_USB_OTG_HS) {
+        /* Disable the ULPI clock in low power mode, this is essential for the
+         * peripheral when using the built-in phy */
+        periph_lpclk_dis(conf->ahb, RCC_AHB1LPENR_OTGHSULPILPEN);
+        /* Only the built-in phy supported for now */
+        assert(conf->phy == STM32_USB_OTG_PHY_BUILTIN);
+        _global_regs(conf)->GUSBCFG |= USB_OTG_GUSBCFG_PHYSEL;
+    }
+#endif
+
+    /* Reset the peripheral after phy selection */
+    _reset_periph(usbdev);
+
+    /* Reset clock */
+    *_pcgcctl_reg(conf) = 0;
+
+    /* Force the peripheral to device mode */
+    _set_mode_device(usbdev);
+
+    /* Disable Vbus detection and force the pull-up on */
+#ifdef STM32_USB_OTG_CID_1x
+    /* Enable no Vbus sensing and enable 'Power Down Disable */
+    _global_regs(usbdev->config)->GCCFG |= USB_OTG_GCCFG_NOVBUSSENS |
+                                           USB_OTG_GCCFG_PWRDWN;
+#else
+    /* Enable no Vbus Detect enable  and enable 'Power Down Disable */
+    _global_regs(usbdev->config)->GCCFG |= USB_OTG_GCCFG_VBDEN |
+                                           USB_OTG_GCCFG_PWRDWN;
+    /* Force Vbus Detect values and ID detect values to device mode */
+    _global_regs(usbdev->config)->GOTGCTL |= USB_OTG_GOTGCTL_VBVALOVAL |
+                                             USB_OTG_GOTGCTL_VBVALOEN |
+                                             USB_OTG_GOTGCTL_BVALOEN |
+                                             USB_OTG_GOTGCTL_BVALOVAL;
+#endif
+    /* disable fancy USB features */
+    _global_regs(conf)->GUSBCFG &=
+        ~(USB_OTG_GUSBCFG_HNPCAP | USB_OTG_GUSBCFG_SRPCAP);
+
+    /* Device mode init */
+    _device_regs(conf)->DCFG |= USB_OTG_DCFG_DSPD_Msk;  /* Full speed! */
+
+    _configure_fifo(usbdev);
+
+    /* Reset the receive FIFO */
+    _flush_rx_fifo(conf);
+
+    /* Reset all TX FIFOs */
+    _flush_fifo(conf, 0x10);
+
+    /* Values from the reference manual tables on TRDT */
+    uint8_t trdt = conf->type == STM32_USB_OTG_FS ? 0x06 : 0x09;
```

Comment:
Is it possible to document this value or add a reference to the doc ?

Reply:
I think I decided against a reference because the exact table number differ between the different MCU reference manuals. I'll try to clarify it a bit